cpu/stm32f1: reworked UART driver

haukepetersen opened the pull request:
now using a simple trick silently introduced with #4840 to put the CPU to sleep at least a little bit during transmission of bytes...

haukepetersen commented:
rebased

@PeterKietzmann: ping

lebrush commented on cpu/stm32f1/periph/uart.c:
```diff
@@ -33,142 +33,109 @@
-        while (!(dev->SR & USART_SR_TXE)) {}
-        dev->DR = data[i];
+        while(!(dev(uart)->SR & USART_SR_TXE)) {
+            cpu_sleep_until_event();
```
Is an event generated when the byte is transmitted?

PeterKietzmann replied:
Can this be affected by #5085?

DipSwitch replied:
It should not be affected apart from working correctly :)

@lebrush Yes an event is generated because the `TXE` and `UE` interrupts are enabled. :)

PeterKietzmann replied:
Nice affect ;-)

lebrush commented:
The code looks good to me :+1:

PeterKietzmann commented on boards/fox/include/periph_conf.h:
```diff
@@ -39,6 +39,9 @@ extern "C" {
 #define CLOCK_APB2_DIV      RCC_CFGR_PPRE2_DIV1     /* APB2 clock -> 72MHz */
 #define CLOCK_APB1_DIV      RCC_CFGR_PPRE1_DIV2     /* APB1 clock -> 36MHz */
+/* resulting bus clocks */
+#define CLOCK_APB1          (CLOCK_CORECLOCK / 2)
```
If I see it correctly, macros in l39-41 aren't used anywhere thus can be removedas well. Having a note for the division by two (like resulting clock frequencies som lines above) should fit.

haukepetersen replied:
not correct -> have a look at the `cpu/stm32f1/cpu.c`...

PeterKietzmann commented on boards/fox/include/periph_conf.h:
```diff
@@ -70,30 +73,30 @@ static const timer_conf_t timer_config[] = {
+        USART2,
+        GPIO_PIN(PORT_A, 3),
+        GPIO_PIN(PORT_A, 2),
+        RCC_APB1ENR_USART2EN,
```
Can you adapt the description in line 77? IMO "bus" is not the correct description. Maybe something like bus-clock-enable or similar?

PeterKietzmann commented on boards/fox/include/periph_conf.h:
```diff
@@ -70,30 +73,30 @@ static const timer_conf_t timer_config[] = {
+        GPIO_PIN(PORT_A, 3),
+        GPIO_PIN(PORT_A, 2),
+        RCC_APB1ENR_USART2EN,
+        APB1,
```
What was rcc bit again :-) ? Maybe spend comment

PeterKietzmann commented on boards/iotlab-m3/include/periph_conf.h:
```diff
@@ -62,8 +65,8 @@ extern "C" {
-    { TIM2, APB1, RCC_APB1ENR_TIM2EN, TIM2_IRQn },
-    { TIM3, APB1, RCC_APB1ENR_TIM3EN, TIM3_IRQn }
+    { TIM2, RCC_APB1ENR_TIM2EN, APB1, TIM2_IRQn },
+    { TIM3, RCC_APB1ENR_TIM3EN, APB1, TIM3_IRQn }
```
I don't understand that change. And besides, why didn't you change it for the other boards?

haukepetersen replied:
Actually, this change should not be here, seems like something went wrong when last rebasing this...

haukepetersen replied:
Now I remember, this actually solved a bug in the timer implementation - but I somehow manged to merge this into this PR. Let me clean this up...

PeterKietzmann commented:
Besides minor comments the PR looks fine. I just tested on an iotlalb node with UART_0. Don't have other (relevant) hardware available

haukepetersen commented:
Rebased and took out the timer fixes (will open another PR for them). But when re-testing, the UART did not work anymore after #5085 was merged. So for now, I removed the call to `cpu_sleep_until_event`. We might want to re-add this once we figure out how to get it to work properly.

DipSwitch commented:
Maybe the pending status must be cleared of TXE and UE.

lebrush commented:
I still don't have a full overview of the difference between events and interrupts... but might it be because the `TXEIE` is not set?

DipSwitch commented:
@lebrush Yup that must be it =)

haukepetersen commented:
that would make sense. I thought that the TX interrupt event (without the TXIE bit actually set) would already count as an event, but this seems not to be true. So maybe this local sleep concept does not work as easily as I thought...

PeterKietzmann commented:
Looks fine for me. UART_0 still works on iotlab-m3, surprise... Would be nice if someone could check if fox or/and the nucleo board still work. Otherwise green lights from my side.

gebart commented:
AFAIK what hardware events that set the SEV event flag is completely up to the CPU vendor to choose, so if it works in one way on nrf5x it won't necessarily be the same concept on stm32.

DipSwitch commented:
@haukepetersen The event interface is only available in the ARM interrupt controller and sleep manager. So for an event to be generated an interrupt line to the NVIC/interrupt controller must set it's pending state.

haukepetersen commented:
@DipSwitch: that makes actually very much sense. 

@PeterKietzmann: I have also no other board than the iotlab here. I think for the `fox` we have no possibility to test at the moment, same goes for the `spark-core`, as mine is broken. Not sure who has access to the `nucleo`... So my bet would be: verify for the `iotlab` and have a good look at the config for the other boards. And as Murdock is happy, I say that should be sufficient...

DipSwitch commented:
How about something like this?

https://github.com/DipSwitch/RIOT/commit/b64f2d58d92d315cbd962fb79f24be38f05b20e8

haukepetersen commented:
@DipSwitch: I think it would work, but then it would make more sense to use DMA right away... I tend to leave it as is for this PR and then we can open a new one improving it.

DipSwitch commented:
ACK

haukepetersen commented:
ALL CIs are happy -> and go.
